File: missionaries-and-cannibals/java/DFS.java
import java.util.*;

public class DFS {
    public static ArrayList<State> DFSearch(State initialState) {
        Map<State, Boolean> visited = new HashMap<State, Boolean>();

        Stack<State> toVisit = new Stack<State>();
        toVisit.push(initialState);

        boolean found = false;

        State next = initialState;

        while(!(toVisit.isEmpty() || found)) {
            next = toVisit.pop();
            System.out.println("popping the stack");

            if (!visited.containsKey(next)) {
                visited.put(next, true);

                if (next.goalTest()) {
                    found = true;
                }
                else {
                    for (State child : next.generateChildren()) {
                        if (!visited.containsKey(child)) {
                            System.out.println("Generating children");
                            System.out.println(child);

                            toVisit.push(child);
                        }
                    }
                }
            }

        }

        return next.backtrack(initialState);

    }
}

File: missionaries-and-cannibals/cs/DFS.cs
using System;
using System.Collections.Generic;

namespace Missionaries
{
    class DFS
    {
        private IState initialState;

        public DFS(IState initialState)
        {
            this.initialState = initialState;
        }
        
        public List<IState> Search()
        {
            Dictionary<IState, bool> visited = new Dictionary<IState, bool>();

            Stack<IState> toVisit = new Stack<IState>();
            toVisit.Push(initialState);

            bool found = false;

            IState currentState;

            while(toVisit.TryPop(out currentState) && !found)
            {
                visited.Add(currentState, true);
                if(currentState.GoalTest())
                    found = true;
                else if(!visited.ContainsKey(currentState))
                {
                    foreach(IState child in currentState.GenerateChildren())
                    {
                        toVisit.Push(child);
                    }
                }
            }
            if (found)
                return currentState.TrackBack(initialState);
            else
                return new List<IState>();
        }
    }
}